File: Frogger/particles.cpp
#include "particles.h"

Particles::Particles(float x, float y, float z) : DynamicObject(x, y, z) {

	setupObjects();
}

Particles::~Particles(){}

void Particles::setupObjects() {
	
	GLfloat v, theta, phi;

	float part_amb[] = { 1.0f, 1.0f, 0.0f, 1.0f };
	float part_diff[] = { 1.0f, 1.0f, 0.0f, 1.0f };
	float part_spec[] = { 1.0f, 1.0f, 0.0f, 1.0f };
	float part_shine[] = { 50.0f };

	for (int i = 0; i < MAX_PARTICULAS; i++){
		
		v = 0.8f*frand() + 0.2f;
		phi = frand()*M_PI;
		theta = 2.0f*frand()*M_PI;

		particulas[i].surface = new VSResSurfRevLib();
		particulas[i].surface->createSphere(0.25f, 25);
		particulas[i].surface->addTexture(0, "assets/particula.bmp");
		particulas[i].surface->setColor(VSResourceLib::AMBIENT, part_amb);
		particulas[i].surface->setColor(VSResourceLib::DIFFUSE, part_diff);
		particulas[i].surface->setColor(VSResourceLib::SPECULAR, part_spec);
		particulas[i].surface->setColor(VSResourceLib::SHININESS, part_shine);
		

		particulas[i].x = position[0];
		particulas[i].y = position[1];
		particulas[i].z = position[2];
		particulas[i].life = 1.0f;
		particulas[i].fade = 0.005f;
		particulas[i].vx = v * cos(theta) * sin(phi);
		particulas[i].vy = v * cos(phi);
		particulas[i].vz = v * sin(theta) * sin(phi);
		particulas[i].ax = 0.1f; /* simular um pouco de vento */
		particulas[i].ay = -0.15f; /* simular a aceleração da gravidade */
		particulas[i].az = 0.0f;
	}
}

void Particles::render(VSShaderLib shader) {
	vsml->loadIdentity(VSMathLib::MODEL);

	for (int i = 0; i < MAX_PARTICULAS; i++){

		if (particulas[i].life > 0.0f) /* só desenha as que ainda estão vivas */
		{
			vsml->pushMatrix(VSMathLib::MODEL);
			vsml->translate(particulas[i].x, particulas[i].y, particulas[i].z);
			particulas[i].surface->render(shader);
			vsml->popMatrix(VSMathLib::MODEL);
		}
	}
}

void Particles::updateParticles() {
	float h;

	h = 0.125f;
	for (int i = 0; i < MAX_PARTICULAS; i++)
	{
		particulas[i].x += (h*particulas[i].vx);
		particulas[i].y -= (h*particulas[i].vy);
		particulas[i].z += (h*particulas[i].vz);
		particulas[i].vx += (h*particulas[i].ax);
		particulas[i].vy += (h*particulas[i].ay);
		particulas[i].vz += (h*particulas[i].az);
		particulas[i].life -= particulas[i].fade;
	}
}

void Particles::resetParticles() {

	for (int i = 0; i < MAX_PARTICULAS; i++)
	{
		GLfloat v, theta, phi;

		v = 0.8f*frand() + 0.2f;
		phi = frand()*M_PI;
		theta = 2.0f*frand()*M_PI;

		particulas[i].x = position[0];
		particulas[i].y = position[1];
		particulas[i].z = position[2];
		particulas[i].life = 1.0f;
		particulas[i].fade = 0.005f;
		particulas[i].vx = v * cos(theta) * sin(phi);
		particulas[i].vy = v * cos(phi);
		particulas[i].vz = v * sin(theta) * sin(phi);
		particulas[i].ax = 0.1f; /* simular um pouco de vento */
		particulas[i].ay = -0.15f; /* simular a aceleração da gravidade */
		particulas[i].az = 0.0f;
	}
}

File: FroggerJS/particles.js
function Particles( x, y, z){

	this.particulas = [];
	
	for(var i = 0; i < 400; i++){//MAX_PARTICLES=400
		this.particulas.push({ sphere: new Sphere(0.25), x: x, y: y, z: z, life: 1, fade: 0.005, vx: 0.0, vy: 0.0, vz: 0.0, ax: 0.1, ay: -0.15, az: 0.0})
	}
	this.setupObjects();
}

Particles.prototype.setupObjects = function(){
	
	var v, theta, phi;

	for (var i = 0; i < 400; i++){ //MAX_PARTICLES=400
		
		v = 0.8*Math.random() + 0.2;
		phi = Math.random()*3.14159265;
		theta = 2.0*Math.random()*3.14159265;

		//this.particulas[i].surface->addTexture(0, "assets/particula.bmp");
		this.particulas[i].sphere.setColor("ambient", [1.0, 1.0, 0.0, 1.0]);
		this.particulas[i].sphere.setColor("diffuse", [1.0, 1.0, 0.0, 1.0]);
		this.particulas[i].sphere.setColor("specular", [1.0, 1.0, 0.0, 1.0]);
		this.particulas[i].sphere.setColor("shininess", 50.0);

		this.particulas[i].vx = v * Math.cos(theta) * Math.sin(phi);
		this.particulas[i].vy = v * Math.cos(phi);
		this.particulas[i].vz = v * Math.sin(theta) * Math.sin(phi);
	}
}

Particles.prototype.render = function(shaderProgram) {
	
	mat4.identity(mMatrix);


	for (var i = 0; i < 400; i++){
		pushMatrix();
		if (this.particulas[i].life > 0.0) /* só desenha as que ainda estão vivas */
		{
			mat4.translate(mMatrix, [this.particulas[i].x, this.particulas[i].y, this.particulas[i].z]);
			this.particulas[i].sphere.render(shaderProgram);
			
		}
		popMatrix();
	}
}

Particles.prototype.updateParticles = function(){
	var h;

	h = 0.125;
	for (var i = 0; i < 400; i++)
	{
		this.particulas[i].x += (h*this.particulas[i].vx);
		this.particulas[i].y -= (h*this.particulas[i].vy);
		this.particulas[i].z += (h*this.particulas[i].vz);
		this.particulas[i].vx += (h*this.particulas[i].ax);
		this.particulas[i].vy += (h*this.particulas[i].ay);
		this.particulas[i].vz += (h*this.particulas[i].az);
		this.particulas[i].life -= this.particulas[i].fade;
	}
}

Particles.prototype.resetParticles = function(){

	for (var i = 0; i < 400; i++)
	{
		var v, theta, phi;

		v = 0.8*Math.random() + 0.2;
		phi = Math.random()*3.14159265;
		theta = 2.0*Math.random()*3.14159265;

		this.particulas[i].x = 35.0;
		this.particulas[i].y = 5.0;
		this.particulas[i].z = 50.0;
		this.particulas[i].life = 1.0;
		this.particulas[i].fade = 0.005;
		this.particulas[i].vx = v * Math.cos(theta) * Math.sin(phi);
		this.particulas[i].vy = v * Math.cos(phi);
		this.particulas[i].vz = v * Math.sin(theta) * Math.sin(phi);
		this.particulas[i].ax = 0.1; /* simular um pouco de vento */
		this.particulas[i].ay = -0.15; /* simular a aceleração da gravidade */
		this.particulas[i].az = 0.0;
	}
}
